src/vm/methodtablebuilder.cpp

```diff
@@ -1522,13 +1522,13 @@ MethodTableBuilder::BuildMethodTableThrowing(
 #if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
             if ((!IsNgenPDBCompilationProcess()
                 && GetAppDomain()->ToCompilationDomain()->GetTargetModule() != g_pObjectClass->GetModule())
-                || (strcmp(className, "Sse") != 0 && strcmp(className, "Sse2") != 0))
+                || strcmp(className, "Avx") == 0 || strcmp(className, "Avx2") == 0 || strcmp(className, "Fma") == 0)
```

Comment:
It would likely be better, long term, if we could check if the class is `Avx` or inherits from it, otherwise I could see this frequently breaking in the future.

This also needs to, currently, explicitly disable BMI1/BMI2 which use the VEX encoding but are not dependent on AVX.
-- AVX requires an OS level opt-in, indicating the 256-bit registers are supported. BMI1/BMI2 also use the VEX encoding, but don't require the OS level opt-in.
-- In some future release, the VEX encoding will likely need to be split out from the AVX support check so that it can be emitted even when `AVX` is not available or disabled.

Reply:
> since AFAIK we don't have an explicit hierarchy.

The hierarchy is exposed in the type system (that is `Avx2` inherits from `Avx`, etc), unless you were referring to something else?

Basically, it would be ideal if we could do a `if ((typeof(T) == typeof(Avx)) || typeof(T).IsSubclassOf(typeof(Avx) || (typeof(T) == typeof(Bmi1)) || (typeof(T) == typeof(Bmi2)))` check. But I'm not sure how feasible that is to do in the method table builder...

Reply:
And regardless of how feasible it is, we should include the `Bmi1` and `Bmi2` classes here, since they also use the VEX encoding.

Reply:
> The hierarchy is exposed in the type system (that is Avx2 inherits from Avx, etc)

Right - I'd forgotten that. Thanks.

Reply:
Sigh, if we can't generate Bmi1/Bmi2 either, this pull request loses pretty much all value.

Here's top 5 intrinsic-related things we were JITting on the PowerShell startup path:


JIT time (ms) | Method |  Intrinsic use
-- | -- | --
3.303 | Utf8Utility.TranscodeToUtf16 | Bmi2
2.45 | Utf8Utility.TranscodeToUtf8 | Bmi2
1.9 | Utf8Utility.GetPointerToFirstInvalidByte | Bmi1
1.708 | Utf16Utility.GetPointerToFirstInvalidChar | `Vector<T>`
0.635 | ASCIIUtility.NarrowUtf16ToAscii | `Vector<T>`

From a 6+ ms win (we could have prejitted top 3) we're down to pretty much nothing.

If it were just up to me, I would close this because the additional complexity doesn't seem justified (as a reminder - this only applies to CoreLib - we don't enable this for user code).

Thoughts?

Reply:
> Sigh, if we can't generate Bmi1/Bmi2 either, this pull request loses pretty much all value.

Why can't we still pregenerate those methods and have the `Bmi1`/`Bmi2` paths be treated as "false" (dead code)? We already are dependent on rejitting to being utilizing the VEX encoding, so I don't see how this is any different.
-- I'm also not sure how this was working if you were already dependent on it. The JIT should be firing an assert if you try to emit a VEX encoded instruction and `canUseVEXEncoding` is not `true` (it returns based on whether or not `Avx` is supported`, which we are setting to false).

As I understand it, this is mostly about improving startup times, not necessarily about maintaining the perf of the generated code (which we can't do, due to the inability for CoreCLR to generate both VEX and non-VEX code paths, currently).

> If it were just up to me, I would close this because the additional complexity doesn't seem justified (as a reminder - this only applies to CoreLib - we don't enable this for user code).

I think it's still a step in the right direction and may still provide benefits to other external consumers of Corelib as well. We likely want all of this code for 5.0, but with some tweaks/design on how it will work for external libraries.

Reply:
> Why can't we still pregenerate those methods and have the Bmi1/Bmi2 paths be treated as "false" (dead code)?

That opens us up to cross-method effects that folks were uncomfortable with:

```csharp
if (Bmi1.IsSupported)
    DoStuffWithBMI();
else
    Fallback();
```

When we up-tier/reJIT the method with the `IsSupported` check, it's going to call into a method that wasn't rejitted yet and will throw exceptions. `ASCIIUtility.GetIndexOfFirstNonAsciiByte` uses the pattern (not for BMI though) so this is not totally academic. GetIndexOfFirstNonAsciiByte_SSE at least asserts the `IsSupported` bit, but that's not enough.

Bugs that show up only when tiering kicks look like "impossible" bugs that only people with tribal knowledge would understand.

Reply:
Isn't that only a problem for methods which use `Avx`, `Avx2`, `Fma`, `Bmi1`, or `Bmi2` without the corresponding check?

Could this be handled by saying any intrinsic in the above cases is emitted as a call? I believe that, ultimately, that would give us:
* If the code is under an `IsSupported` check, it will be treated as "dead code" for the purposes of AOT and be dropped.
* If the code is not under an `IsSupported` check, it will be a call to the appropriate intrinsic method (which will be JIT'd)

It would make the `DoStuffWithBMI` method slower, until it is rejitted, but should allow the above scenario to work.
-- A "proper" fix would probably be to treat methods which use `Avx`, `Avx2`, `Fma`, `Bmi1`, or `Bmi2` without an appropriate `IsSupported` check as "must be jitted", but that is likely a more complex change.

Reply:
It's unfortunate that this is so complex, and that the "right answer" is elusive, but I think we're converging.

I think @tannergooding's suggestion is in the right direction.

> When we up-tier/reJIT the method with the IsSupported check, it's going to call into a method that wasn't rejitted yet and will throw exceptions. 

As @tannergooding points out, this is only an issue if code lacks the corresponding check. That's why we're only doing this for SPC.dll at present - because we're confident that it has the necessary checks. Otherwise, we'd have bugs on older systems. Any method that uses the intrinsics without the check would not be AOT-compiled, but presumably would not be invoked on the start-up path because the start-up path will all be assuming no AVX.

> A "proper" fix would probably be to treat methods which use Avx, Avx2, Fma, Bmi1, or Bmi2 without an appropriate IsSupported check as "must be jitted", but that is likely a more complex change.

Looking further out, I wonder whether we could recognize and AOT compile methods that are known to only be invoked on certain platforms (i.e. the guard is in the calling method). It would mean that the VM would have to invoke the compiler with different options for those methods, and we'd have to rely on SPC.dll to always have the right guards in place before invoking such methods. But, it wouldn't require the JIT to have to support mixed encodings in a single method.

But that is likely an even more complex change.

Reply:
Making the BMI intrinsics into normal calls (outside IsSupported) is a good trick! Fixed.